src/mds/MDSRank.cc

```diff
@@ -36,11 +36,244 @@
+  }
+
+  void clear_mdlog() {
+    Context *ctx = new FunctionContext([this](int r) {
+        handle_clear_mdlog(r);
+      });
+    mdlog->wait_for_safe(new MDSInternalContextWrapper(mdsrank, ctx));
+  }
+
+  void handle_clear_mdlog(int r) {
+    if (r != 0) {
+      *ss << "Error " << r << " (" << cpp_strerror(r) << ") while flushing journal";
+      complete(r);
+      return;
+    }
+
+    trim_mdlog();
+  }
+
+  void trim_mdlog() {
+    int ret = mdlog->trim_all();
+    if (ret != 0) {
+      *ss << "Error " << ret << " (" << cpp_strerror(ret) << ") while trimming log";
+      complete(ret);
+      return;
+    }
+
+    expire_segments();
+  }
+
+  void expire_segments() {
+    MDSGatherBuilder *expiry_gather = new MDSGatherBuilder(g_ceph_context);
+
+    const auto expiring_segments = mdlog->get_expiring_segments();
+    for (auto p : expiring_segments) {
+      p->wait_for_expiry(expiry_gather->new_sub());
+    }
+
+    if (!expiry_gather->has_subs()) {
+      trim_expired_segments();
+      return;
+    }
+
+    Context *ctx = new C_OnFinisher(
+      new FunctionContext([this](int r) {
+          handle_expire_segments(r);
+        }), mdsrank->finisher);
+    expiry_gather->set_finisher(new MDSInternalContextWrapper(mdsrank, ctx));
+    expiry_gather->activate();
+  }
+
+  void handle_expire_segments(int r) {
+    if (r != 0) {
+      complete(r);
+      return;
+    }
+
+    trim_expired_segments();
+  }
+
+  void trim_expired_segments() {
+    mdlog->trim_expired_segments();
+    write_journal_head();
+  }
+
+  void write_journal_head() {
+    mdlog->get_journaler()->write_head(this);
+  }
+
+  void finish(int r) override {
+    on_finish->complete(r);
+  }
+
+  MDCache *mdcache;
+  MDLog *mdlog;
+  MDSRank *mdsrank;
+  std::ostream *ss;
+  Context *on_finish;
+};
+
+class C_Drop_Cache : public Context {
+public:
+  C_Drop_Cache(Server *server, MDCache *mdcache,
+               MDLog *mdlog, MDSRank *mdsrank,
+               int64_t recall_timeout, Formatter *f, Context *on_finish)
+    : server(server), mdcache(mdcache), mdlog(mdlog), mdsrank(mdsrank),
+      recall_timeout(recall_timeout), f(f), on_finish(on_finish) {
+  }
+
+  void send() {
+    // not really a hard requirement here, but lets ensure this in
+    // case we change the logic here.
+    assert(mdsrank->mds_lock.is_locked());
+
+    if (recall_timeout < 0) {
+      cmd_err(f, "invalid timeout value -- should be >= 0");
+      complete(-EINVAL);
+      return;
+    }
+
+    recall_client_state();
+  }
+
+private:
+
+  void recall_client_state() {
+    f->open_object_section("result");
+
+    MDSGatherBuilder gather(g_ceph_context);
+    server->recall_client_state(1.0, true, &gather);
+    if (gather.has_subs()) {
+      MDSContextSleeper *sleeper = new MDSContextSleeper("cache_drop_waiter", mdsrank);
+      gather.set_finisher(sleeper);
+      gather.activate();
+
+      mdsrank->mds_lock.Unlock();
+      retval = sleeper->wait_for(recall_timeout);
```

Comment:
We don't want to block the thread at all here. Let's use a `SafeTimer` (? maybe the right choice) which runs its finisher after `recall_timeout`. I think this will require two detachable contexts (like `MDSContextSleeper`). One waits for all clients to trim caps and the other waits for the Timer to complete. We won't be using a condition variable here because we don't want to wait at all. (`MDSContextSleeper` may get axed since I don't think we will use it.)

Reply:
Looking at this again, specifically it will block the finisher thread which is a problem. (The dispatcher thread is fortunately free now, which is good!)